### ISSUE TITLE
Replacing the LDCMD hack

### ISSUE THREAD

**Issue by levitte** [openssl/openssl#4935]
Back just a week or so ago, it was possible to link stuff using a different command than your usual compiler, through the LDCMD environment / make variable.  WIth the removal of `Makefile.shared` also removed that possibility.

This is unfortunate for Google's fuzzing initiative, as shown in [their issue on this](https://github.com/google/oss-fuzz/issues/1037).

*One* way is to [retrofit the hack back into the unixly Makefile template](https://github.com/google/oss-fuzz/issues/1037#issuecomment-351845889)...  that hack is currently incomplete, as it only affects the linking of applications, not of shared libraries, but I do believe that for the particular use case, that is enough.

However, I'm not exactly happy with this hack, so I'd like to see if there are thoughts on how to do this more cleanly.  One idea I have is to add these lines in `Configurations/unix-Makefile.tmpl`:

``` make
LD=$(CC)
LDXX=$(CXX)
```

... and then use them appropriately in the perl functions `lib2obj2shlib` and `obj2bin` instead of `$(CC)` and `$(CXX)`.  I imagine that would enable anyone who wants to to do, say, `make LD='$(CXX)'`...

Note: we cannot ignore this issue, because [fuzz/README.md](https://github.com/openssl/openssl/blob/master/fuzz/README.md).  It can be argued that it's mostly useful for *us*, and that an ugly hack might therefore be acceptable, but frankly, I'd rather not leave something that raises eyebrows a few years down the line...

**Comment by dot-asm** [comment]
I'd say that LD would probably be suboptimal choice. Because it's defined by default by make and hence people might have some [rather specific] expectations about what impact it should have. Well, not that it's impossible to use it reliably in situation in question, it's rather a usability thing. Though I can imagine it causing problems if OpenSSL build procedure is "wrapped" into some bigger build procedure [that sets up LD for another component sake]...

**Comment by dot-asm** [comment]
> LD would probably be suboptimal choice

I meant suboptimal choice for the *name* of variable.

**Comment by levitte** [comment]
Note that we already defined CC and CXX, so we already have *that* particular problem...  in the middle of a meeting, so not enough brains to think this through, let med get back on the matter later (don't let me stop you from commenting further, though).

**Comment by dot-asm** [comment]
> Note that we already defined CC and CXX, so we already have that particular problem...

Not quite. As mentioned, if you view it as *usability* problem, us defining CC and CXX is not a problem, because we don't break common expectations [that $(CC) is C compiler and that it accepts the "usual" C flags]. While suggestion here is to define $(LD) and pass $(CFLAGS) flags. This is a bit counter-intuitive when you take into consideration that default $(LD) value is `ld`. But once again, it's not like it would be impossible to make it work and work reliably [in standalone build], it's just that choice of the name might be viewed as *suboptimal*.

**Comment by levitte** [comment]
Ah, I do get that point.  Do you have a better name idea then?

**Comment by dot-asm** [comment]
I don't know... If one has to avoid LDCMD (to minimize risk of misunderstanding), then how about CC_LINK, C_LINK, LINK_C, CC_LD, LD_CC? And CXX_LINK, LINK_CXX, etc. correspondingly... No strong preference as long as it doesn't overload some default value...

**Comment by levitte** [comment]
Actually, I think that we do need to use the LDCMD env variable, for the sake of build compatibility within the 1.1.x series.  *But*, that doesn't stop us from doing the cleaner thing as well.  I think I'll go for `CC_LINK` and `CXX_LINK`

**Comment by levitte** [comment]
I'd like to take up this again, as part of the discussion on flags.  It's glaringly apparent that we're locking out some cases where linking isn't practical or appropriate to do with `$(CC)`.

Having looked around a bit (among others at what automake does), there seems to be a precedent with the make variables `CCLD` and `CXXLD`, and I think it might appropriate to do the same.  Their values would, of course, default to `$(CC)` and `$(CXX)`, but may be something else if the config target has a different idea, or the user give `make` different values.

This also means that we need to be more careful exactly how we give those commands their flags.  `$(CFLAGS)` isn't necessarely an appropriate set of flags...

**Comment by levitte** [comment]
Note that at this point in time, I'm unsure if it's appropriate to do this work for 1.1.1, or if it should be post-1.1.1.  The current milestone is 1.1.1 but can easily be changed.  Opinions, anyone?

**Comment by kroeckx** [comment]
It sort of sounds like a new feature to me.

**Comment by mattcaswell** [comment]
Ping @levitte. What now with this?

**Comment by levitte** [comment]
I'm pushed this post 1.1.1.  Whatever we do with this for 1.1.1 will probably be another temporary hack, and I think that the config parameters and build templates are due for an overhaul, which would involve fixing this issue as well, but with consistency.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
